File: MATLAB/push_relabel.m
function [ max_flow, F, height ] = push_relabel( C, n)
INF = 20;
F = zeros(n);
 excess = zeros(1,n);
  height = zeros(1,n);
  seen = zeros(1,n);
  list = zeros(1,n-2);
  
  for i=1:n-2
      list(i) = i+1;
  end
  
  height(1) = n;
  excess(1) = INF;
  for i=1:n
    [F,excess] = push(C, F, excess, 1, i);
  end
  p = 1;
  while (p < n - 1)
     u = list(p);
     old_height = height(u);
     [ F,excess,seen,height ] = discharge(C, F, excess, height, seen, u, n, INF);
     if (height(u) > old_height)
       list = move_to_front(p,list);
       p = 1;
     else
       p = p + 1;
     end
  end
  max_flow = 0;
   for i=1:n
     max_flow = max_flow + F(1,i);
   end
end

File: C++/push_relabel.cpp
#include <stdlib.h>

#define MIN(X,Y) ((X) < (Y) ? (X) : (Y))
#define INFINITE 10000000

void push(const int * const * C, int ** F, int *excess, int u, int v);

void relabel(const int * const * C, const int * const * F, int *height, int u, int nodes) {
  int v;
  int min_height = INFINITE;
  for (v = 0; v < nodes; v++) {
    if (C[u][v] - F[u][v] > 0) {
      min_height = MIN(min_height, height[v]);
      height[u] = min_height + 1;
    }
  }
}

void discharge(const int * const * C, int ** F, int *excess, int *height, int *seen, int u, int nodes) {
  while (excess[u] > 0) {
    if (seen[u] < nodes) {
      int v = seen[u];
      if ((C[u][v] - F[u][v] > 0) && (height[u] > height[v])){
    push(C, F, excess, u, v);
      }
      else
    seen[u] += 1;
    } else {
      relabel(C, F, height, u, nodes);
      seen[u] = 0;
    }
  }
}

void moveToFront(int i, int *A) {
  int temp = A[i];
  int n;
  for (n = i; n > 0; n--){
    A[n] = A[n-1];
  }
  A[0] = temp;
}

int pushRelabel(const int * const * C, int ** F, int source, int sink, int nodes) {
  int *excess, *height, *list, *seen, i, p;

  excess = (int *) calloc(nodes, sizeof(int));
  height = (int *) calloc(nodes, sizeof(int));
  seen = (int *) calloc(nodes, sizeof(int));

  list = (int *) calloc((nodes-2), sizeof(int));

  for (i = 0, p = 0; i < nodes; i++){
    if((i != source) && (i != sink)) {
      list[p] = i;
      p++;
    }
  }

  height[source] = nodes;
  excess[source] = INFINITE;
  for (i = 0; i < nodes; i++)
    push(C, F, excess, source, i);
  p = 0;
   while (p < nodes - 2) {
     int u = list[p];
     int old_height = height[u];
     discharge(C, F, excess, height, seen, u, nodes);
     if (height[u] > old_height) {
       moveToFront(p,list);
       p=0;
     }
     else
       p += 1;
   }
   int maxflow = 0;
   for (i = 0; i < nodes; i++)
     maxflow += F[source][i];

   free(list);

   free(seen);
   free(height);
   free(excess);

   return maxflow;
 }
